<html><head></head><body>
        <section>

                            <header>
                    <h1 class="header-title">Errors and Logging</h1>
                </header>
            
            <article>
                
<p class="mce-root">In this chapter, we will learn how to deal with errors and return an error when you need to. Go's error mechanism is different to those of some of the other popular languages, and this section will teach you how to work with errors the Go way. We will also learn how to perform simple logging operations in your application to get insights into your running application for better debugging. We will cover the following topics in this chapter:</p>
<ul>
<li>Creating custom error types</li>
<li>The <kbd>try...catch</kbd> equivalent in Go</li>
<li>Doing simple logging in your app</li>
<li>Gracefully dealing with panics</li>
</ul>


            </article>

            
        </section>
    

        <section>

                            <header>
                    <h1 class="header-title">Creating custom error types</h1>
                </header>
            
            <article>
                
<p class="mce-root">Let's start with creating your custom error type. If you come from languages such as C# and Java, you may find the error mechanism a little different in Go. Moreover, the way you create your own custom error is very simple because Go is a duck-typed language, which means that you are good to go as long as your struct satisfies an interface. Let's go ahead and create our own custom error using a new type. So, I will have two fields, <kbd>ShortMessage</kbd> and the <kbd>DetailedMessage</kbd> of string type. You can have as many fields as you want, to capture more information about your errors. Furthermore, to satisfy the <kbd>error</kbd> interface, I'm going to implement a new method, <kbd>*MyError</kbd>, which will return a <kbd>string</kbd> value, and we can output this error to either our console or some log file.</p>
<p class="mce-root"/>
<p class="mce-root">Then, what I'm going to do is to return the error message. So, the way you do this is very simple: you can just return this error type from your method. Let's imagine that we have a <kbd>doSomething</kbd> method that returns an error. Let's imagine we did some lines of codes in that method and it returns an error for some reason, such as a <kbd>ShortMessage</kbd> instance of <kbd>"Wohoo something happened!"</kbd>. Of course, you will probably need to use more meaningful messages here, and don't forget to use this <kbd>&amp;</kbd> operator. It will get the address of your <kbd>*MyError</kbd> object, since we are working with a pointer here. If you don't do this, you will see there's a type error, and one way to fix is this is to just remove that <kbd>*</kbd> pointer and the error will be will fixed. But you probably don't want to have multiple copies of the same object, so instead of doing what I have just described, you can easily do this: send a reference back so that you have better memory management. Let's look at the whole code now:</p>
<pre>package main<br/><br/>import "fmt"<br/><br/>type MyError struct{<br/>  ShortMessage string<br/>  DetailedMessage string<br/>  //Name string<br/>  //Age int<br/>}<br/><br/>func (e *MyError) Error() string {<br/>  return e.ShortMessage + "\n" +e.DetailedMessage<br/><br/>}<br/>  func main(){<br/>    err:= doSomething()<br/>    fmt.Print(err)<br/>}<br/>func doSomething() error {<br/>  //Doing something here...<br/>  return &amp;MyError{ShortMessage:"Wohoo something happened!", DetailedMessage:"File cannot found!"}<br/>}</pre>
<p class="mce-root"/>
<p class="mce-root">So, let's run this, and of course it will return some errors; we're just going to add <kbd>err</kbd> here and then run it to the console. Now, we can see that our message or error message is written to the console, which is shown in the following screenshot:</p>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-954 image-border" src="assets/b6df9a90-ea0c-4398-817c-d7addf448f5d.png" style="width:26.00em;height:15.17em;"/></p>
<p class="mce-root">That's how you can simply create your own error message types. In our next section, we're going to learn the <kbd>try...catch</kbd> equivalent in Go.</p>


            </article>

            
        </section>
    

        <section>

                            <header>
                    <h1 class="header-title">The try...catch equivalent in Go</h1>
                </header>
            
            <article>
                
<p>Unlike in other languages, there is no <kbd>try...catch</kbd> block in Go. In this section, we're going to see how Go handles basic errors. So, the first thing we're going to see is how to handle the errors returned by an API calls. We can use the <kbd>time.Parse()</kbd> method for that as it accepts a layout and a value string. It returns two things, one is the <kbd>parsedDate</kbd> and the other one is an error. Instead of returning an exception, Go returns an error as its second parameter most of the time.</p>
<p class="mce-root"/>
<p class="mce-root"/>
<p>Now, the way you can handle this is to check whether the <kbd>parsedDate</kbd> is nil. If it's not nil in Go, then we know that an error has happened and we need to handle it. <span>If nothing happens, we can safely proceed to our next line, </span><span>and that is to write the content of our <kbd>parsedDate</kbd> to output. So, for this, check the following code example:</span></p>
<pre class="CDPAlignLeft CDPAlign">package main<br/><br/>import (<br/>  "time"<br/>  "fmt"<br/>)<br/><br/>func main(){<br/>  parsedDate, err:= time.Parse("2006", "2018")<br/>  if err != nil {<br/>    fmt.Println("An error occured", err.Error())<br/>  }else{<br/>    fmt.Println(parsedDate)<br/>  }<br/>}</pre>
<p><span>The preceding code will give you the following output:</span></p>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-955 image-border" src="assets/150d4ada-4907-439b-b23e-bcb639aa084a.png" style="width:26.50em;height:15.00em;"/></p>
<p>You can see that it works fine. What happens if we add some <kbd>string</kbd> values after <kbd>2018</kbd>? Let's add <kbd>abc</kbd>, and run the code. If you see the following screenshot, you can see that an error occurred in parsing time; it also added the error message <kbd>An error occured parsing time "2018 abc": extra text: abc</kbd>, as shown in the following screenshot:</p>
<p class="mce-root"/>
<p class="mce-root"/>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-956 image-border" src="assets/9f6e2e3d-0e2b-4cef-99b2-2e1070ab3f0c.png" style="width:63.33em;height:36.58em;"/></p>
<p>Now, the second part of this section is when you return an error yourself. Let's say we have the <kbd>doSomething</kbd> function and it returns an <kbd>err</kbd> type. Check the following code:</p>
<pre>package main<br/>import (<br/>  "fmt"<br/>  "errors"<br/>)<br/>func main(){<br/>  _, err := doSomething()<br/>  if err != nil {<br/>    fmt.Println(err)<br/>  }<br/>}<br/>func doSomething() (string,error) {<br/>  return "", errors.New("Something happened.")<br/>}</pre>
<p class="mce-root"/>
<p class="mce-root"/>
<p><span>The preceding code will give the following output:</span></p>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-957 image-border" src="assets/9fdffdf3-e102-4817-b4a6-6db1eb04ef37.png" style="width:26.92em;height:15.75em;"/></p>
<p><span>So, that's how you can do a simple <kbd>try...catch</kbd> equivalent in Go. In our next section, we're going to see how to do simple logging in your application.</span></p>


            </article>

            
        </section>
    

        <section>

                            <header>
                    <h1 class="header-title">Doing simple logging in your application</h1>
                </header>
            
            <article>
                
<p>In this section, we are going to learn how you can do simple logging in your application. Of course, there are various ways you can do this, and there are also third-party packages that allow you to do it as well, but we're going to use the <kbd>log</kbd> package that Go provides. So, the first thing we have to do is create a new file by using the <kbd>os</kbd> package, and if, somehow, when creating the <kbd>log</kbd> file, an error happens, we're going to write it to the console. We will also be using the <kbd>defer</kbd> function. Before the <kbd>main</kbd> method exits this <kbd>defer</kbd> function will be called, and the next step will be to set the output:</p>
<pre>package main<br/>import (<br/>  "os"<br/>  "fmt"<br/>  "log"<br/>)<br/>func main(){<br/>  log_file, err := os.Create("log_file")<br/>  if err != nil{<br/>    fmt.Println("An error occured...")<br/>  }<br/>  defer log_file.Close()<br/>  log.SetOutput(log_file)</pre>
<pre><br/>  log.Println("Doing some logging here...")<br/>  log.Fatalln("Fatal: Application crashed!")<br/>}</pre>
<p>When we run the preceding code, a new file called <kbd>log_file</kbd> is created with the following content:</p>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-958 image-border" src="assets/e907f02e-c236-4823-a6a7-946bd6ab4355.png" style="width:36.58em;height:7.50em;"/></p>
<p>You may wonder what the difference is between a fatal error and a normal info error. Let's reorder the two lines and see how that new order behaves. Hence, we will run <kbd>Fatalln</kbd> first and then <kbd>Println</kbd> as follows:</p>
<pre>package main<br/>import (<br/>  "os"<br/>  "fmt"<br/>  "log"<br/>)<br/>func main(){<br/>  log_file, err := os.Create("log_file")<br/>  if err != nil{<br/>    fmt.Println("An error occured...")<br/>  }<br/>  defer log_file.Close()<br/>  log.SetOutput(log_file)<br/>  log.Fatalln("Fatal: Application crashed!")<br/>  log.Println("Doing some logging here...")<br/>}</pre>
<p>If you now run the preceding code and check the content of the <kbd>log_file</kbd>, you will see that the second <kbd>Println</kbd> did not get written:</p>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-959 image-border" src="assets/a37bdce9-51b4-4c7c-9ead-f353274673bc.png" style="width:33.50em;height:6.58em;"/></p>
<p>The difference is that <kbd>Fatalln</kbd> is similar to <kbd>Println</kbd> but only followed by a call to <kbd>os.Exit</kbd>. So, it basically writes a log and exits out of the application, which is the simple difference between the two. This is how you can simply do logging in your application. Of course, you can wrap <kbd>main</kbd> function into your package if you don't want to be setting the output all the time as we did here. In our next section, we're going to see how to gracefully deal with panics.</p>


            </article>

            
        </section>
    

        <section>

                            <header>
                    <h1 class="header-title">Gracefully dealing with panics</h1>
                </header>
            
            <article>
                
<p>In this section, we're going to see how to gracefully deal with panics. Unlike with errors, if you don't recover from panics, it will stop the execution of your program. Therefore, dealing with them is important if you want your program to continue. First of all, let's see how we can throw a panic in our Go program. You can simply use a keyword called <kbd>panic</kbd>, which is a built-in function, type panicked, and run it to obtain the output:</p>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-960 image-border" src="assets/c8bfa192-2f73-42ac-886d-0b6f49bb6bd6.png" style="width:61.50em;height:25.25em;"/></p>
<p class="mce-root"/>
<p class="mce-root"/>
<p>There's another way to do this. Let's use another function here and write something. Let's imagine we're doing something and for some reason it just panicked. This might be a third-party method, which means it is located in a third-party package, so we may not have full control over that package. So, here, if you run the preceding code, this is what we're going to see in the application window, along with the message we want to write to the console, which is as follows:</p>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-961 image-border" src="assets/a142bb04-f555-4a89-a8f0-c9a7eec7fa69.png" style="width:61.25em;height:30.58em;"/></p>
<p>We also see the stack trace of our <kbd>panic</kbd> here. First, it hit the main message, and later, it hit the <kbd>writeSomething()</kbd> method. So how can we deal with this <kbd>panic</kbd>? We have this <kbd>defer</kbd> keyword and you have to use this <kbd>defer</kbd>. The <kbd>defer</kbd> means that; well, let's say that right before your method exits, you want to run another piece of code, so you just pass a function and you say "I want to run this <kbd>defer</kbd> function". Of course, it needs to be like this: <kbd>defer func(){}()</kbd>, or you can just say <kbd>defer writeSomething()</kbd> here. It doesn't matter, but since I'm going to run a couple of codes, I encapsulated them within function here. We have another keyword called <kbd>recover</kbd> that runs the <kbd>defer</kbd> function right before this <kbd>main</kbd> function exits. Also, within this function, we try <kbd>recover</kbd>.</p>
<p class="mce-root"/>
<p class="mce-root"/>
<p>If a panic happened, this <kbd>recover</kbd> returns something, and if there's no panic, that means it will not return something. Therefore, the value of <kbd>r</kbd> will be <kbd>nil</kbd> , which means we just don't write anything to the console because we don't need to. However, if the panic happened, then we'll just enter the <kbd>if</kbd> condition and we'll just write whatever comes from the <kbd>recover</kbd> building method and just go ahead and run the following code, and we will get the corresponding output:</p>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-962 image-border" src="assets/e79fd976-228d-41d9-be1e-1c15e47e6d58.png" style="width:30.42em;height:40.50em;"/></p>
<p>So, now you can see that we basically say <kbd>Recovered in f</kbd> and the message is whatever the panic threw, and that's something we wrote here. If you want to see the continuation of this process, we can just copy the <kbd>defer func()</kbd> function from the <kbd>main</kbd> function. Next, we will create another method called <kbd>sayHello()</kbd> and paste the <kbd>defer func()</kbd> into it. What I'm trying to show you is that we have recovered from the panic, so the execution will also hit this line as well. So, we can just go ahead and run the following code:</p>
<pre>package main<br/><br/>import "fmt"<br/><br/>func main(){<br/>  sayHello()<br/>  fmt.Println("After the panic was recovered!")<br/>}<br/><br/>func sayHello(){<br/>  defer func(){<br/>    if r := recover(); r != nil {<br/>      fmt.Println("Recovered in f", r)<br/>    }<br/>  }()<br/>  writeSomething()<br/>}<br/><br/>func writeSomething(){<br/>  /// Doing something here..<br/>  panic("Write operation error")<br/>}</pre>
<p>After the execution <kbd>main</kbd> function, so now we see the message:</p>
<p class="CDPAlignCenter CDPAlign"><img class="alignnone size-full wp-image-963 image-border" src="assets/f04575d2-ccdd-4269-94d3-ee71ba297f3b.png" style="width:29.08em;height:15.33em;"/></p>
<p class="mce-root"/>
<p class="mce-root"/>
<p>If we didn't have the <kbd>defer</kbd> function, let's see how that would have behaved. Now you see that it didn't hit <kbd>main</kbd> function, we have the panic and all the stack trace here, and that's how you can gracefully handle panics in your application. </p>


            </article>

            
        </section>
    

        <section>

                            <header>
                    <h1 class="header-title">Summary</h1>
                </header>
            
            <article>
                
<p>This chapter was a walkthrough of errors and logging. In the next chapter, <span>we will learn how to work with files and directories in an operating system. We will also learn about parsing and using various formats, such as XML, YAML, and JSON.</span></p>


            </article>

            
        </section>
    </body></html>